pkg/tags/tag.go

```diff
@@ -0,0 +1,276 @@
+	Name      string        // a name tag for this tag
+	Address   swarm.Address // the associated swarm hash for this tag
+	StartedAt time.Time     // tag started to calculate ETA
+
+	// end-to-end tag tracing
+	ctx      context.Context  // tracing context
+	span     opentracing.Span // tracing root span
+	spanOnce sync.Once        // make sure we close root span only once
+}
+
+// NewTag creates a new tag, and returns it
+func NewTag(uid uint32, s string, total int64, anon bool) *Tag {
+	t := &Tag{
+		Uid:       uid,
+		Anonymous: anon,
+		Name:      s,
+		StartedAt: time.Now(),
+		Total:     total,
+	}
+
+	// context here is used only to store the root span `new.upload.tag` within Tag,
+	// we don't need any type of ctx Deadline or cancellation for this particular ctx
+	t.ctx, t.span = spancontext.StartSpan(context.Background(), "new.upload.tag")
+	return t
+}
+
+// Context accessor
+func (t *Tag) Context() context.Context {
+	return t.ctx
+}
+
+// FinishRootSpan closes the pushsync span of the tags
+func (t *Tag) FinishRootSpan() {
+	t.spanOnce.Do(func() {
+		t.span.Finish()
+	})
+}
+
+// IncN increments the count for a state
+func (t *Tag) IncN(state State, n int) {
+	var v *int64
+	switch state {
+	case StateSplit:
+		v = &t.Split
+	case StateStored:
+		v = &t.Stored
+	case StateSeen:
+		v = &t.Seen
+	case StateSent:
+		v = &t.Sent
+	case StateSynced:
+		v = &t.Synced
+	}
+	atomic.AddInt64(v, int64(n))
+}
+
+// Inc increments the count for a state
+func (t *Tag) Inc(state State) {
+	t.IncN(state, 1)
+}
+
+// Get returns the count for a state on a tag
+func (t *Tag) Get(state State) int64 {
+	var v *int64
+	switch state {
+	case StateSplit:
+		v = &t.Split
+	case StateStored:
+		v = &t.Stored
+	case StateSeen:
+		v = &t.Seen
+	case StateSent:
+		v = &t.Sent
+	case StateSynced:
+		v = &t.Synced
+	}
+	return atomic.LoadInt64(v)
+}
+
+// GetTotal returns the total count
+func (t *Tag) TotalCounter() int64 {
+	return atomic.LoadInt64(&t.Total)
+}
+
+// WaitTillDone returns without error once the tag is complete
+// wrt the state given as argument
+// it returns an error if the context is done
+func (t *Tag) WaitTillDone(ctx context.Context, s State) error {
+	if t.Done(s) {
+		return nil
+	}
+	ticker := time.NewTicker(100 * time.Millisecond)
+	for {
+		select {
+		case <-ticker.C:
+			if t.Done(s) {
+				return nil
+			}
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+// Done returns true if tag is complete wrt the state given as argument
+func (t *Tag) Done(s State) bool {
+	n, total, err := t.Status(s)
+	return err == nil && n == total
+}
+
+// DoneSplit sets total count to SPLIT count and sets the associated swarm hash for this tag
+// is meant to be called when splitter finishes for input streams of unknown size
+func (t *Tag) DoneSplit(address swarm.Address) int64 {
+	total := atomic.LoadInt64(&t.Split)
+	atomic.StoreInt64(&t.Total, total)
+	t.Address = address
```

Comment:
i wonder if we should copy the address here, it could result in quirks. since address is a struct that holds a slice, this infers a copy of the slice pointer. wdyt?

Reply:
Yes, this function is copied as it is from the swarm. We should address this in a separate PR.